universal/include/userver/formats/universal/universal.hpp

```diff
@@ -0,0 +1,322 @@
+    builder[std::string(boost::pfr::get_name<I, T>())] = Value;
+  };
+};
+
+template <typename T, auto I, typename Value, typename Field, auto DefaultValue>
+constexpr inline auto RunParseCheckFor(const Value&, std::optional<Field>& field, Default<DefaultValue>) {
+  if(!field.has_value()) {
+    field = DefaultValue;
+  };
+};
+
+template <typename T, auto I, typename... Params, typename Builder, typename Value>
+constexpr inline std::enable_if_t<utils::AnyOf(utils::IsSameCarried<Additional>(), utils::TypeList<Params...>{}), void>
+RunWrite(Builder& builder, const std::unordered_map<std::string, Value>& field) {
+  for(const auto& element : field) {
+    builder[element.first] = element.second;
+  };
+};
+
+template <typename T, auto I, typename Builder, typename Field>
+constexpr inline auto RunCheckFor(Builder&, Field&&, detail::Disabled) noexcept {};
+
+template <typename T, auto I, typename Builder, typename Field>
+constexpr inline auto RunCheckFor(Builder&, const std::optional<Field>&, detail::Disabled) noexcept {};
+
+template <typename T, auto I, typename Builder, typename Field, typename CheckT>
+constexpr inline auto RunCheckFor(Builder&, Field&& field, CheckT check) {
+  if(!Check(field, check)) {
+    throw std::runtime_error(ErrorMessage<T, I>(std::forward<Field>(field), check));
+  };
+};
+
+} // namespace exam
+template <typename T, auto I, typename Builder, typename... Params>
+constexpr inline auto UniversalSerializeField(
+     FieldParametries<T, I, Params...>
+    ,Builder& builder
+    ,const T& obj) {
+  const auto& value = boost::pfr::get<I>(obj);
+  using exam::RunCheckFor;
+  using exam::RunWrite;
+  (RunCheckFor<T, I>(builder, value, Params{}), ...);
+  RunWrite<T, I, Params...>(builder, value);
+};
+
+template <typename T, auto I, typename Format, typename... Params>
+constexpr inline auto UniversalParseField(
+     FieldParametries<T, I, Params...>
+    ,Format&& from) {
+  using exam::Read;
+  using exam::RunParseCheckFor;
+  using FieldType = std::remove_cvref_t<decltype(boost::pfr::get<I>(std::declval<T>()))>;
+  auto value = Read<T, I, Params...>(from, userver::formats::parse::To<FieldType>{});
+  (RunParseCheckFor<T, I>(from, value, Params{}), ...);
+  return value;
+};
+
+template <typename T, auto I, typename Format, typename... Params>
+constexpr inline std::optional<std::remove_cvref_t<decltype(boost::pfr::get<I>(std::declval<T>()))>>
+UniversalTryParseField(
+     FieldParametries<T, I, Params...>
+    ,Format&& from) noexcept {
+  using FieldType = std::remove_cvref_t<decltype(boost::pfr::get<I>(std::declval<T>()))>;
+  using exam::Check;
+  using exam::Read;
+
+
+  auto val = Read<T, I, Params...>(from, userver::formats::parse::To<std::optional<FieldType>>{});
+  if(val && (Check(*val, Params{}) && ...)) {
+    return val;
+  };
+  return std::nullopt;
+};
+
+} // namespace detail
+
+template <typename T, typename... Params>
+class SerializationConfig {
+  private:
+    template <auto I, typename Param>
```

Comment:
Order of class members:

```
- public:
   - types
   - constructors and factory functions
   - special member functions
   - methods
   - data members
- protected: ...
- private: ...
```

This should be applied to other classes in the PR, too, like `ConstexprString`

Reply:
This still applies to `ConstexprString`